File: app/src/main/java/com/coinninja/coinkeeper/service/client/SignedCoinKeeperApiClient.java
package com.coinninja.coinkeeper.service.client;

import androidx.annotation.NonNull;

import com.coinninja.bindings.model.Transaction;
import com.coinninja.coinkeeper.bitcoin.BroadcastProvider;
import com.coinninja.coinkeeper.bitcoin.BroadcastingClient;
import com.coinninja.coinkeeper.model.Contact;
import com.coinninja.coinkeeper.model.db.DropbitMeIdentity;
import com.coinninja.coinkeeper.service.client.model.CNDevice;
import com.coinninja.coinkeeper.service.client.model.CNPhoneNumber;
import com.coinninja.coinkeeper.service.client.model.CNSharedMemo;
import com.coinninja.coinkeeper.service.client.model.CNTopicSubscription;
import com.coinninja.coinkeeper.service.client.model.CNUserPatch;
import com.coinninja.coinkeeper.service.client.model.InviteUserPayload;
import com.coinninja.coinkeeper.service.client.model.NewsArticle;
import com.coinninja.coinkeeper.ui.market.Granularity;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import retrofit2.Response;

import static com.coinninja.coinkeeper.util.DropbitIntents.CN_API_CREATE_DEVICE_APPLICATION_KEY;
import static com.coinninja.coinkeeper.util.DropbitIntents.CN_API_CREATE_DEVICE_APPLICATION_NAME;
import static com.coinninja.coinkeeper.util.DropbitIntents.CN_API_CREATE_DEVICE_PLATFORM_KEY;
import static com.coinninja.coinkeeper.util.DropbitIntents.CN_API_CREATE_DEVICE_UUID_KEY;

public class SignedCoinKeeperApiClient extends CoinKeeperApiClient implements BroadcastingClient {

    private final String fcmAppId;

    public SignedCoinKeeperApiClient(CoinKeeperClient client, String fcmAppId) {
        super(client);
        this.fcmAppId = fcmAppId;
    }

    public Response resendVerification(CNPhoneNumber CNPhoneNumber) {
        return executeCall(getClient().resendVerification(CNPhoneNumber));
    }

    public Response registerUserAccount(CNPhoneNumber CNPhoneNumber) {
        return executeCall(getClient().createUserAccount(CNPhoneNumber));
    }

    public Response registerWallet(String publicVerificationKey) {
        JsonObject json = new JsonObject();
        json.addProperty("public_key_string", publicVerificationKey);
        return executeCall(getClient().createWallet(json));
    }

    public Response fetchContactStatus(List<Contact> contacts) {
        JsonArray numbers = new JsonArray();

        for (Contact contact : contacts) {
            numbers.add(contact.getHash());
        }

        JsonObject query = createQuery("phone_number_hash", numbers);

        return executeCall(getClient().queryUsers(query));
    }

    public Response getCNWalletAddresses() {
        return executeCall(getClient().getWalletAddresses());
    }

    public Response addAddress(String address) {
        JsonObject json = new JsonObject();
        json.addProperty("address", address);
        return executeCall(getClient().addAddressToCNWallet(json));
    }

    public Response addAddress(String address, String publicKey) {
        JsonObject json = new JsonObject();
        json.addProperty("address", address);
        json.addProperty("address_pubkey", publicKey);
        return executeCall(getClient().addAddressToCNWallet(json));
    }

    public Response queryWalletAddress(String phoneHash) {
        JsonArray numbers = new JsonArray();
        numbers.add(phoneHash);
        JsonObject request = createQuery("phone_number_hash", numbers);
        JsonObject query = request.getAsJsonObject("query");
        query.addProperty("address_pubkey", true);
        return executeCall(getClient().queryWalletAddress(request));
    }

    public Response inviteUser(InviteUserPayload inviteUserPayload) {
        return executeCall(getClient().inviteUser(inviteUserPayload));
    }

    public Response getReceivedInvites() {
        return executeCall(getClient().getAllIncomingInvites());
    }

    public Response getSentInvites() {
        return executeCall(getClient().getAllSentInvites());
    }

    public Response sendAddressForInvite(String inviteId, String btcAddress, String addressPubKey) {
        JsonObject query = new JsonObject();

        query.addProperty("wallet_address_request_id", inviteId);
        query.addProperty("address", btcAddress);
        query.addProperty("address_pubkey", addressPubKey);

        return executeCall(getClient().sendAddress(query));
    }

    public Response patchSuppressionForWalletAddressRequest(String inviteId) {
        JsonObject query = new JsonObject();

        query.addProperty("suppress", false);
        return executeCall(getClient().patchInvite(inviteId, query));
    }

    public Response updateInviteStatusCompleted(String inviteId, String txID) {
        JsonObject query = new JsonObject();

        query.addProperty("status", "completed");
        query.addProperty("txid", txID);

        return executeCall(getClient().patchInvite(inviteId, query));
    }

    public Response updateInviteStatusCanceled(String inviteId) {
        JsonObject query = new JsonObject();

        query.addProperty("status", "canceled");
        return executeCall(getClient().patchInvite(inviteId, query));
    }

    public Response removeAddress(String address) {
        return executeCall(getClient().removeAddress(address));
    }

    public Response resetWallet() {
        return executeCall(getClient().resetWallet());
    }

    public Response createUpdateInviteStatusError(String inviteServerID, String errorMessage) {
        JsonObject query = new JsonObject();

        query.addProperty("status", "canceled");
        return createTeaPotErrorFor(getClient().patchInvite(inviteServerID, query), errorMessage);
    }

    public Response verifyWallet() {
        return executeCall(getClient().verifyWallet());
    }

    public Response verifyAccount() {
        return executeCall(getClient().verifyUserAccount());
    }

    public Response getCNMessages(CNElasticSearch elasticSearch) {
        return executeCall(getClient().getCNMessages(elasticSearch.toJson()));
    }

    public Response getTransactionNotification(String txid) {
        return executeCall(getClient().getTransactionNotification(txid));
    }

    public Response createCNDevice(String uuid) {
        String applicationName = CN_API_CREATE_DEVICE_APPLICATION_NAME;
        JsonObject query = new JsonObject();

        query.addProperty(CN_API_CREATE_DEVICE_APPLICATION_KEY, applicationName);
        query.addProperty(CN_API_CREATE_DEVICE_PLATFORM_KEY, CNDevice.DevicePlatform.ANDROID.toString());
        query.addProperty(CN_API_CREATE_DEVICE_UUID_KEY, uuid);

        return executeCall(getClient().createCNDevice(query));
    }

    public Response registerForPushEndpoint(String cnDeviceId, String token) {
        JsonObject queryBody = new JsonObject();

        queryBody.addProperty("application", fcmAppId);
        queryBody.addProperty("platform", "GCM");
        queryBody.addProperty("token", token);

        return executeCall(getClient().registerForPushEndpoint(cnDeviceId, queryBody));
    }

    public Response fetchRemoteEndpointsFor(String cnDeviceId) {

        return executeCall(getClient().fetchRemoteEndpointsFor(cnDeviceId));
    }

    public Response unRegisterDeviceEndpoint(String cnDeviceId, String endpoint) {
        return executeCall(getClient().unRegisterDeviceEndpoint(cnDeviceId, endpoint));
    }

    public Response fetchDeviceEndpointSubscriptions(String devicesId, String deviceEndpoint) {
        return executeCall(getClient().fetchDeviceEndpointSubscriptions(devicesId, deviceEndpoint));
    }

    public Response subscribeToTopics(String deviceId, String endpointId, CNTopicSubscription topics) {
        return executeCall(getClient().subscribeToTopics(deviceId, endpointId, topics));
    }

    public Response subscribeToWalletNotifications(String deviceEndpoint) {
        JsonObject jsonObject = buildWalletSubscriptionBody(deviceEndpoint);
        return executeCall(getClient().subscribeToWalletTopic(jsonObject));
    }

    public Response updateWalletSubscription(String deviceEndpoint) {
        JsonObject jsonObject = buildWalletSubscriptionBody(deviceEndpoint);
        return executeCall(getClient().updateWalletSubscription(jsonObject));
    }

    public Response postTransactionNotification(CNSharedMemo sharedMemo) {
        return executeCall(getClient().postTransactionNotification(sharedMemo));
    }

    public Response disableDropBitMeAccount() {
        return executeCall(getClient().patchUserAccount(new CNUserPatch(true)));
    }

    public Response enableDropBitMeAccount() {
        return executeCall(getClient().patchUserAccount(new CNUserPatch(false)));
    }

    @NotNull
    public Response getIdentity(@Nullable DropbitMeIdentity identity) {
        return executeCall(getClient().getIdentity(identity.getServerId()));
    }

    @NotNull
    public Response deleteIdentity(@Nullable DropbitMeIdentity identity) {
        return executeCall(getClient().deleteIdentity(identity.getServerId()));
    }

    public Response getIdentities() {
        return executeCall(getClient().getIdentities());
    }

    @NotNull
    public Response addIdentity(@Nullable CNUserIdentity identity) {
        return executeCall(getClient().addIdentity(identity));
    }

    @NotNull
    public Response verifyIdentity(@Nullable CNUserIdentity identity) {
        return executeCall(getClient().verifyIdentity(identity));
    }

    @NotNull
    public Response createUserFromIdentity(@Nullable CNUserIdentity identity) {
        return executeCall(getClient().createUserFrom(identity));
    }

    @NotNull
    public Response unsubscribeFromTopic(@NotNull String devicesId, @NotNull String deviceEndpoint, @NotNull String topicId) {
        return executeCall(getClient().unsubscribeFromTopic(devicesId, deviceEndpoint, topicId));
    }

    public Response loadHistoricPricing(@NotNull Granularity granularity) {
        return executeCall(getClient().loadHistoricPricing(granularity.getValue()));
    }

    @SuppressWarnings("unchecked")
    public Response loadNews(int count, int offset) {
        count = offset > 0 ? count + offset : count;
        Response response = executeCall(getClient().loadNews(count));
        if (response.code() == 200 && offset > 0) {
            List<NewsArticle> fetched = (List<NewsArticle>) response.body();
            if (fetched != null && fetched.size() == count) {
                ArrayList articles = new ArrayList(fetched) {
                    public ArrayList trimOffset() {
                        removeRange(0, offset);
                        return this;
                    }
                }.trimOffset();
                response = Response.success(articles);
            }
        }

        return response;
    }

    @Override
    public Response<Object> broadcastTransaction(@NotNull Transaction transaction) {
        RequestBody requestBody = RequestBody.create(MediaType.parse("text/plain"), transaction.rawTx);
        return executeCall(getClient().broadcastTransaction(requestBody));
    }

    @NotNull
    @Override
    public BroadcastProvider broadcastProvider() {
        return BroadcastProvider.COIN_NINJA;
    }

    @NonNull
    private JsonObject buildWalletSubscriptionBody(String deviceEndpoint) {
        JsonObject jsonObject = new JsonObject();
        jsonObject.addProperty("device_endpoint_id", deviceEndpoint);
        return jsonObject;
    }

}

File: app/src/main/java/com/coinninja/coinkeeper/service/client/SignedCoinKeeperApiClient.kt
package com.coinninja.coinkeeper.service.client

import app.coinninja.cn.libbitcoin.model.Transaction
import app.dropbit.annotations.Mockable
import app.dropbit.commons.util.removeRange
import com.coinninja.coinkeeper.bitcoin.BroadcastProvider
import com.coinninja.coinkeeper.bitcoin.BroadcastingClient
import com.coinninja.coinkeeper.model.Contact
import com.coinninja.coinkeeper.model.db.DropbitMeIdentity
import com.coinninja.coinkeeper.service.client.model.*
import com.coinninja.coinkeeper.ui.market.Granularity
import com.coinninja.coinkeeper.util.DropbitIntents.*
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import okhttp3.MediaType
import okhttp3.RequestBody
import retrofit2.Response

@Mockable
@Suppress("UNCHECKED_CAST")
class SignedCoinKeeperApiClient(
        client: CoinKeeperClient, private val fcmAppId: String
) : CoinKeeperApiClient(client), BroadcastingClient {

    val cnWalletAddresses: Response<List<CNWalletAddress>>
        get() = executeCall(client.walletAddresses)

    val receivedInvites: Response<List<ReceivedInvite>>
        get() = executeCall(client.allIncomingInvites)

    val sentInvites: Response<List<SentInvite>>
        get() = executeCall(client.allSentInvites)

    val identities: Response<List<CNUserIdentity>>
        get() = executeCall(client.identities)

    fun resendVerification(CNPhoneNumber: CNPhoneNumber): Response<CNUserAccount> {
        return executeCall(client.resendVerification(CNPhoneNumber))
    }

    fun registerUserAccount(CNPhoneNumber: CNPhoneNumber): Response<CNUserAccount> {
        return executeCall(client.createUserAccount(CNPhoneNumber))
    }

    fun registerWallet(walletRegistrationPayload: WalletRegistrationPayload): Response<CNWallet> {
        return executeCall(client.createWallet(walletRegistrationPayload))
    }

    fun fetchContactStatus(contacts: List<Contact>): Response<Map<String, String>> {
        val numbers = JsonArray()

        for (contact in contacts) {
            numbers.add(contact.hash)
        }

        val query = createQuery("phone_number_hash", numbers)

        return executeCall(client.queryUsers(query))
    }

    fun addAddress(address: String, publicKey: String): Response<CNWalletAddress> {
        val json = JsonObject()
        json.addProperty("address", address)
        json.addProperty("address_pubkey", publicKey)
        return executeCall(client.addAddressToCNWallet(json))
    }

    fun queryWalletAddress(phoneHash: String): Response<List<AddressLookupResult>> {
        val numbers = JsonArray()
        numbers.add(phoneHash)
        val request = createQuery("phone_number_hash", numbers)
        val query = request.getAsJsonObject("query")
        query.addProperty("address_pubkey", true)
        return executeCall(client.queryWalletAddress(request))
    }

    fun inviteUser(inviteUserPayload: InviteUserPayload): Response<InvitedContact> {
        return executeCall(client.inviteUser(inviteUserPayload))
    }

    fun sendAddressForInvite(inviteId: String, btcAddress: String, addressPubKey: String): Response<CNWalletAddress> {
        val query = JsonObject()

        query.addProperty("wallet_address_request_id", inviteId)
        query.addProperty("address", btcAddress)
        query.addProperty("address_pubkey", addressPubKey)

        return executeCall(client.sendAddress(query))
    }

    fun patchSuppressionForWalletAddressRequest(inviteId: String): Response<SentInvite> {
        val query = JsonObject()

        query.addProperty("suppress", false)
        return executeCall(client.patchInvite(inviteId, query))
    }

    fun updateInviteStatusCompleted(inviteId: String, txID: String): Response<SentInvite> {
        val query = JsonObject()

        query.addProperty("status", "completed")
        query.addProperty("txid", txID)

        return executeCall(client.patchInvite(inviteId, query))
    }

    fun updateInviteStatusCanceled(inviteId: String): Response<SentInvite> {
        val query = JsonObject()

        query.addProperty("status", "canceled")
        return executeCall(client.patchInvite(inviteId, query))
    }

    fun removeAddress(address: String): Response<Map<String, String>> {
        return executeCall(client.removeAddress(address))
    }

    fun resetWallet(): Response<Void> {
        return executeCall(client.resetWallet())
    }

    fun createUpdateInviteStatusError(inviteServerID: String, errorMessage: String): Response<SentInvite> {
        val query = JsonObject()

        query.addProperty("status", "canceled")
        return createTeaPotErrorFor(client.patchInvite(inviteServerID, query), errorMessage)
    }

    fun verifyWallet(): Response<CNWallet> {
        return executeCall(client.verifyWallet())
    }

    fun disableWallet(): Response<CNWallet> {
        return executeCall(client.disableWallet(DisableWalletRequest()))
    }

    fun replaceWalletWith(replaceWalletRequest: ReplaceWalletRequest): Response<CNWallet> {
        return executeCall(client.replaceWalletWith(replaceWalletRequest))
    }

    fun verifyAccount(): Response<CNUserAccount> {
        return executeCall(client.verifyUserAccount())
    }

    fun getCNMessages(elasticSearch: CNElasticSearch): Response<List<CNGlobalMessage>> {
        return executeCall(client.getCNMessages(elasticSearch.toJson()))
    }

    fun getTransactionNotification(txid: String): Response<List<CNSharedMemo>> {
        return executeCall(client.getTransactionNotification(txid))
    }

    fun createCNDevice(uuid: String): Response<CNDevice> {
        val applicationName = CN_API_CREATE_DEVICE_APPLICATION_NAME
        val query = JsonObject()

        query.addProperty(CN_API_CREATE_DEVICE_APPLICATION_KEY, applicationName)
        query.addProperty(CN_API_CREATE_DEVICE_PLATFORM_KEY, CNDevice.DevicePlatform.ANDROID.toString())
        query.addProperty(CN_API_CREATE_DEVICE_UUID_KEY, uuid)

        return executeCall(client.createCNDevice(query))
    }

    fun registerForPushEndpoint(cnDeviceId: String, token: String): Response<CNDeviceEndpoint> {
        val queryBody = JsonObject()

        queryBody.addProperty("application", fcmAppId)
        queryBody.addProperty("platform", "GCM")
        queryBody.addProperty("token", token)

        return executeCall(client.registerForPushEndpoint(cnDeviceId, queryBody))
    }

    fun fetchRemoteEndpointsFor(cnDeviceId: String): Response<List<CNDeviceEndpoint>> {

        return executeCall(client.fetchRemoteEndpointsFor(cnDeviceId))
    }

    fun unRegisterDeviceEndpoint(cnDeviceId: String, endpoint: String): Response<Map<String, String>> {
        return executeCall(client.unRegisterDeviceEndpoint(cnDeviceId, endpoint))
    }

    fun fetchDeviceEndpointSubscriptions(devicesId: String, deviceEndpoint: String): Response<CNSubscriptionState> {
        return executeCall(client.fetchDeviceEndpointSubscriptions(devicesId, deviceEndpoint))
    }

    fun subscribeToTopics(deviceId: String, endpointId: String, topics: CNTopicSubscription): Response<Void> {
        return executeCall(client.subscribeToTopics(deviceId, endpointId, topics))
    }

    fun subscribeToWalletNotifications(deviceEndpoint: String): Response<CNSubscription> {
        val jsonObject = buildWalletSubscriptionBody(deviceEndpoint)
        return executeCall(client.subscribeToWalletTopic(jsonObject))
    }

    fun updateWalletSubscription(deviceEndpoint: String): Response<CNSubscription> {
        val jsonObject = buildWalletSubscriptionBody(deviceEndpoint)
        return executeCall(client.updateWalletSubscription(jsonObject))
    }

    fun postTransactionNotification(sharedMemo: CNSharedMemo): Response<CNTransactionNotificationResponse> {
        return executeCall(client.postTransactionNotification(sharedMemo))
    }

    fun disableDropBitMeAccount(): Response<CNUserPatch> {
        return executeCall(client.patchUserAccount(CNUserPatch(true)))
    }

    fun enableDropBitMeAccount(): Response<CNUserPatch> {
        return executeCall(client.patchUserAccount(CNUserPatch(false)))
    }

    fun getIdentity(identity: DropbitMeIdentity?): Response<*> {
        return executeCall(client.getIdentity(identity!!.serverId))
    }

    fun deleteIdentity(identity: DropbitMeIdentity?): Response<Void> {
        return executeCall(client.deleteIdentity(identity!!.serverId))
    }

    fun addIdentity(identity: CNUserIdentity): Response<CNUserIdentity> {
        return executeCall(client.addIdentity(identity))
    }

    fun verifyIdentity(identity: CNUserIdentity): Response<CNUserAccount> {
        return executeCall(client.verifyIdentity(identity))
    }

    fun createUserFromIdentity(identity: CNUserIdentity): Response<CNUserAccount> {
        return executeCall(client.createUserFrom(identity))
    }

    fun unsubscribeFromTopic(devicesId: String, deviceEndpoint: String, topicId: String): Response<Void> {
        return executeCall(client.unsubscribeFromTopic(devicesId, deviceEndpoint, topicId))
    }

    fun loadHistoricPricing(granularity: Granularity): Response<List<HistoricalPriceRecord>> {
        return executeCall(client.loadHistoricPricing(granularity.value))
    }

    fun loadNews(numArticles: Int, offset: Int): Response<MutableList<NewsArticle>> {
        val count = if (offset > 0) numArticles + offset else numArticles
        var response: Response<MutableList<NewsArticle>> = executeCall(client.loadNews(count))
        if (response.code() == 200 && offset > 0) {
            val fetched = response.body() as MutableList<NewsArticle>
            if (fetched.size == count) {
                fetched.removeRange(0 until offset)
                response = Response.success(fetched)
            }
        }

        return response
    }

    override fun broadcastTransaction(transaction: Transaction): Response<Any> {
        val requestBody = RequestBody.create(MediaType.parse("text/plain"), transaction.encodedTransaction)
        return executeCall(client.broadcastTransaction(requestBody))
    }

    override fun broadcastProvider(): BroadcastProvider {
        return BroadcastProvider.COIN_NINJA
    }

    private fun buildWalletSubscriptionBody(deviceEndpoint: String): JsonObject {
        val jsonObject = JsonObject()
        jsonObject.addProperty("device_endpoint_id", deviceEndpoint)
        return jsonObject
    }


}
